src/participation.js

```diff
@@ -0,0 +1,197 @@
+      var ins = [
+        UI.rdf.st(subject, ns.wf('participation'), participation, padDoc),
+
+        UI.rdf.st(participation, ns.wf('participant'), me, padDoc),
+        UI.rdf.st(participation, ns.cal('dtstart'), new Date(), padDoc),
+        UI.rdf.st(
+          participation,
+          ns.ui('backgroundColor'),
+          UI.pad.lightColorHash(me),
+          padDoc
+        )
+      ]
+      kb.updater.update([], ins, function (uri, ok, errorMessage) {
+        if (!ok) {
+          reject(new Error('Error recording your partipation: ' + errorMessage))
+        } else {
+          resolve(participation)
+        }
+      })
+      resolve(participation)
+    }
+  })
+}
+
+/** Record my participation and display participants
+ *
+ * @param {NamedNode} subject - the thing in which participation is happening
+ * @param {NamedNode} padDoc - The document into which the particpation should be recorded
+ * @param {DOMNode} refreshable - A DOM element whose refresh() is to be called if the change works
+ *
+ */
+UI.participationrecordParticipation = function (subject, padDoc, refreshable) {
+  var me = UI.authn.currentUser()
+  if (!me) return // Not logged in
+
+  var parps = kb.each(subject, ns.wf('participation')).filter(function (pn) {
+    return kb.holds(pn, ns.wf('participant'), me)
+  })
+  if (parps.length > 1) {
+    throw new Error('Multiple records of your participation')
+  }
+  if (parps.length) {
+    // If I am not already recorded
+    return parps[0] // returns the particpation object
+  } else {
+    var participation = UI.widgets.newThing(padDoc)
+    var ins = [
+      UI.rdf.st(subject, ns.wf('participation'), participation, padDoc),
+
+      UI.rdf.st(participation, ns.wf('participant'), me, padDoc),
+      UI.rdf.st(participation, UI.ns.cal('dtstart'), new Date(), padDoc),
+      UI.rdf.st(
+        participation,
+        ns.ui('backgroundColor'),
+        UI.pad.lightColorHash(me),
+        padDoc
+      )
+    ]
+    kb.updater.update([], ins, function (uri, ok, errorMessage) {
+      if (!ok) {
+        throw new Error('Error recording your partipation: ' + errorMessage)
+      }
+      if (refreshable && refreshable.refresh) {
+        refreshable.refresh()
+      }
+      // UI.pad.renderPartipants(dom, table, padDoc, subject, me, options)
+    })
+    return participation
+  }
+}
+
+// Record my participation and display participants
+//
+participationModule.manageParticipation = function (
+  dom,
+  container,
+  padDoc,
+  subject,
+  me,
+  options
+) {
+  var table = dom.createElement('table')
+  container.appendChild(table)
+  participationModule.renderPartipants(dom, table, padDoc, subject, me, options)
+  try {
+    participationModule.recordParticipation(subject, padDoc, table)
+  } catch (e) {
+    container.appendChild(
+      UI.widgets.errorMessageBlock(
+        dom,
+        'Error recording your partipation: ' + e
+      )
+    ) // Clean up?
+  }
+  return table
+}
```

Comment:
This whole commit seems to be about moving participation out of src/pad.js and into its own src/participation.js

In itself that looks like a good thing! But unrelated to notepad copy, so I'll move this into a separate PR. Also, in the new team we established a rule-of-thumb that we move any code we touch to typescript, and i would also like to add docs, unit tests, and examples in the process.

Reply:
If you want I can take a look at this, as I was working on pad.js testing?  And there were also some complexity warnings that I had either resolved or was working on.  Maybe I can combine these into one change.

Reply:
If you're offering, I won't say no! :) https://github.com/solid/solid-ui/issues/210#event-3056026823